vendor/bat-native-ledger/src/bat/ledger/internal/uphold/uphold_authorization_unittest.cc

```diff
@@ -0,0 +1,300 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <tuple>
+#include <utility>
+
+#include "base/test/task_environment.h"
+#include "bat/ledger/internal/database/database_mock.h"
+#include "bat/ledger/internal/ledger_client_mock.h"
+#include "bat/ledger/internal/ledger_impl_mock.h"
+#include "bat/ledger/internal/state/state_keys.h"
+#include "bat/ledger/internal/uphold/uphold_authorization.h"
+#include "bat/ledger/option_keys.h"
+#include "net/http/http_status_code.h"
+
+// npm run test -- brave_unit_tests --filter=UpholdAuthorizationTest*
+
+using ::testing::_;
+using ::testing::Invoke;
+using ::testing::TestParamInfo;
+using ::testing::TestWithParam;
+using ::testing::Values;
+
+namespace ledger {
+namespace uphold {
+
+using GetAnonFundsTestParamType =
+    std::tuple<bool,               // fetch old balance enabled
+               std::string,        // Brave wallet
+               type::UrlResponse,  // balance server response
+               std::pair<type::Result,
+                         base::Optional<double>>  // expected result
+               >;
+
+class GetAnonFundsTest : public TestWithParam<GetAnonFundsTestParamType> {
+ private:
+  base::test::TaskEnvironment scoped_task_environment_;
+
+ protected:
+  std::unique_ptr<MockLedgerClient> mock_ledger_client_;
+  std::unique_ptr<MockLedgerImpl> mock_ledger_impl_;
+  std::unique_ptr<database::MockDatabase> mock_database_;
+  std::unique_ptr<UpholdAuthorization> authorization_;
+
+ public:
+  GetAnonFundsTest()
+      : mock_ledger_client_{std::make_unique<MockLedgerClient>()},
+        mock_ledger_impl_{
+            std::make_unique<MockLedgerImpl>(mock_ledger_client_.get())},
+        mock_database_{
+            std::make_unique<database::MockDatabase>(mock_ledger_impl_.get())},
+        authorization_{
+            std::make_unique<UpholdAuthorization>(mock_ledger_impl_.get())} {}
+};
+
+std::string GetAnonFundsNameSuffixGenerator(
+    const TestParamInfo<GetAnonFundsTest::ParamType>& info) {
+  const bool fetch_old_balance_enabled = std::get<0>(info.param);
+  const std::string& brave_wallet = std::get<1>(info.param);
+  const type::UrlResponse& balance_server_response = std::get<2>(info.param);
+
+  if (!fetch_old_balance_enabled) {
+    return "FetchOldBalanceDisabled";
+  }
+
+  if (brave_wallet.empty()) {
+    return "BraveWalletIsNotCreated";
+  }
+
+  if (brave_wallet.find(R"("payment_id": "")") != std::string::npos) {
+    return "BraveWalletPaymentIdIsEmpty";
+  }
+
+  if (balance_server_response.status_code != net::HttpStatusCode::HTTP_OK) {
+    return "BalanceServerError";
+  }
+
+  if (balance_server_response.body.empty()) {
+    return "InvalidBodyInBalanceServerResponse";
+  }
+
+  return "HappyPath";
+}
+
+INSTANTIATE_TEST_SUITE_P(

```

Comment:
Nice find w/ parameterized tests! These are really cool

For other reviewers, more info at https://source.chromium.org/chromium/chromium/src/+/main:third_party/googletest/src/googletest/include/gtest/gtest-param-test.h;l=71-93;drc=0a3a3a845e136a9a6ccd8e9b924b848840f22b7b